crates/biome_js_analyze/src/lint/nursery/no_irregular_whitespace.rs

```diff
@@ -0,0 +1,112 @@
+];
+
+declare_lint_rule! {
+    /// Disallows the use of irregular whitespace characters.
+    ///
+    /// Invalid or irregular whitespace causes issues with ECMAScript 5 parsers and also makes code harder to debug.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// constcount=1;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = 'thing';
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const count = 1;
+    /// ```
+    ///
+    /// ```js
+    /// const foo = '';
+    /// ```
+    ///
+    pub NoIrregularWhitespace {
+        version: "next",
+        name: "noIrregularWhitespace",
+        language: "js",
+        recommended: false,
+        sources: &[RuleSource::Eslint("no-irregular-whitespace")],
+        source_kind: RuleSourceKind::SameLogic,
+    }
+}
+
+impl Rule for NoIrregularWhitespace {
+    type Query = Ast<JsModule>;
+    type State = TextRange;
+    type Signals = Vec<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        get_irregular_whitespace(node)
+    }
+
+    fn diagnostic(_ctx: &RuleContext<Self>, range: &Self::State) -> Option<RuleDiagnostic> {
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                range,
+                markup! {
+                    "Irregular whitespaces found."
+                },
+            )
+            .note(markup! {
+                "Replace the irregular whitespaces with normal whitespaces or tabs."
+            }),
+        )
+    }
+}
+
+fn get_irregular_whitespace(node: &JsModule) -> Vec<TextRange> {
+    let syntax = node.syntax();
+    let mut all_whitespaces_trivia: Vec<SyntaxTriviaPiece<JsLanguage>> = vec![];
+    let is_whitespace = |trivia: &SyntaxTriviaPiece<JsLanguage>| {
+        trivia.is_whitespace() && !trivia.text().replace(' ', "").is_empty()
+    };
+
+    for token in syntax.descendants_tokens(Direction::Next) {
+        let leading_trivia_pieces = token.leading_trivia().pieces();
+        let trailing_trivia_pieces = token.trailing_trivia().pieces();
+
+        for trivia in leading_trivia_pieces {
+            if is_whitespace(&trivia) {
+                all_whitespaces_trivia.push(trivia);
+            }
+        }
+
+        for trivia in trailing_trivia_pieces {
+            if is_whitespace(&trivia) {
+                all_whitespaces_trivia.push(trivia);
+            }
+        }
+    }
+
+    all_whitespaces_trivia
+        .iter()
+        .filter_map(|trivia| {
+            let has_irregular_whitespace = trivia.text().chars().any(|char| {
+                IRREGULAR_WHITESPACES
+                    .iter()
+                    .any(|irregular_whitespace| &char == irregular_whitespace)
+            });
+            has_irregular_whitespace.then(|| trivia.text_range())
+        })
+        .collect::<Vec<TextRange>>()
```

Comment:
Can you try a little experiment, and see if a double for loop can perform a bit better? We create an iterator every time we inspect a character, this might be expensive.

Reply:
sure! I will give it a try

Reply:
Now that I think about it, it won't change a thing. Sorry about that

Reply:
oh, it's okay